rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+                    r = listener_6.notified() => r?,
+                }
+
+                let resolvers = crate::client::resolvers::get()?;
+                tracing::info!(?resolvers);
+            }
+
+            listener_4.close()?;
+            listener_6.close()?;
+            Ok::<_, anyhow::Error>(())
+        })?;
+
+        Ok(())
+    }
+
+    pub(crate) struct CombinedListener {
+        listener_4: Listener,
+        listener_6: Listener,
+    }
+
+    impl CombinedListener {
+        pub(crate) fn new() -> Result<Self> {
+            let [key_ipv4, key_ipv6] = open_network_registry_keys()?;
+            let listener_4 = Listener::new(key_ipv4)?;
+            let listener_6 = Listener::new(key_ipv6)?;
+
+            Ok(Self {
+                listener_4,
+                listener_6,
+            })
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            self.listener_4.close()?;
+            self.listener_6.close()?;
+            tracing::debug!("Unregistered registry listener");
+            Ok(())
+        }
+
+        pub(crate) async fn notified(&mut self) -> Result<()> {
+            tokio::select! {
+                r = self.listener_4.notified() => r?,
+                r = self.listener_6.notified() => r?,
+            }
+            Ok(())
+        }
+    }
+
+    struct Listener {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
+        wait_handle: Option<HANDLE>,
+        event: Option<HANDLE>,
+    }
+
+    impl Listener {
+        pub(crate) fn new(key: winreg::RegKey) -> Result<Self> {
+            let notify = Box::pin(Notify::default());
+            let notify_ptr: *const Notify = notify.deref();
+            let event = unsafe { CreateEventA(None, false, false, None) }?;
+            let mut wait_handle = HANDLE(0isize);
+
+            // The docs say that `RegisterWaitForSingleObject` uses "a worker thread" from
+            // "the thread pool".
+            // Raymond Chen, who is an authority on Windows internals, says RegisterWaitForSingleObject
+            // does multiple waits on the same worker thread internally:
+            // <https://devblogs.microsoft.com/oldnewthing/20081117-00/?p=20183>
+            // There is a different function in the CLR called `RegisterWaitForSingleObject`,
+            // so he might be talking about that.
+            // If he's not, then we're invisibly tying up two worker threads. Can't help it.
+
+            // SAFETY: It's complicated.
+            // The callback here can cause a lot of problems. We pin the `Notify` object.
+            // We don't use an `Arc` because `Notify` already has only `&self` methods, and
+            // the callback has no way to free an `Arc`.
+            // When we call `UnregisterWaitEx` later, we wait for all callbacks to finish running
+            // before we drop the `Notify`, to avoid a dangling pointer.
+            unsafe {
+                RegisterWaitForSingleObject(
```

Comment:
just to make sure I understand this `RegisterWaitForSingleObject` will signal `wait_handle`, every time `event` is signaled, even when the event that `event` is listening for changes?